include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -2055,6 +2055,202 @@ __parallel_reduce_by_segment(oneapi::dpl::__internal::__device_backend_tag, _Exe
+// parallel_scan_by_segment - sync pattern
+//------------------------------------------------------------------------
+template <typename _CustomName, bool __is_inclusive, typename _Range1, typename _Range2, typename _Range3,
+          typename _BinaryPredicate, typename _BinaryOperator, typename _InitType>
+__future<sycl::event, __result_and_scratch_storage<
+                          oneapi::dpl::__internal::tuple<std::uint32_t, oneapi::dpl::__internal::__value_t<_Range2>>>>
+__parallel_scan_by_segment_reduce_then_scan(sycl::queue& __q, _Range1&& __keys, _Range2&& __values,
+                                            _Range3&& __out_values, _BinaryPredicate __binary_pred,
+                                            _BinaryOperator __binary_op, [[maybe_unused]] _InitType __init)
+{
+    using _GenReduceInput = __gen_scan_by_seg_reduce_input<_BinaryPredicate>;
+    using _ReduceOp = __scan_by_seg_op<_BinaryOperator>;
+    using _GenScanInput = __gen_scan_by_seg_scan_input<_BinaryPredicate>;
+    using _ScanInputTransform = __get_zeroth_element;
+    using _ValueType = oneapi::dpl::__internal::__value_t<_Range2>;
+    const std::size_t __n = __keys.size();
+    // TODO: A bool type may be used here for a smaller footprint in registers / temp storage but results in IGC crashes
+    // during JIT time. The same occurs for uint8_t and uint16_t. uint32_t is used as a workaround until the underlying
+    // issue is resolved.
+    using _FlagType = std::uint32_t;
+    using _PackedFlagValueType = oneapi::dpl::__internal::tuple<_FlagType, _ValueType>;
+    using _WrappedInitType =
+        std::conditional_t<__is_inclusive, oneapi::dpl::unseq_backend::__no_init_value<_PackedFlagValueType>,
+                           oneapi::dpl::unseq_backend::__init_value<_PackedFlagValueType>>;
+    if constexpr (__is_inclusive)
+    {
+        _WrappedInitType __wrapped_init;
+        using _WriteOp = __write_scan_by_seg<__is_inclusive, _WrappedInitType, _BinaryOperator>;
+        return __parallel_transform_reduce_then_scan<sizeof(_PackedFlagValueType), _CustomName>(
+            __q, __n,
+            oneapi::dpl::__ranges::make_zip_view(std::forward<_Range1>(__keys), std::forward<_Range2>(__values)),
+            std::forward<_Range3>(__out_values), _GenReduceInput{__binary_pred}, _ReduceOp{__binary_op},
+            _GenScanInput{}, _ScanInputTransform{}, _WriteOp{__wrapped_init, __binary_op}, __wrapped_init,
+            /*Inclusive*/ std::true_type{}, /*_IsUniquePattern=*/std::false_type{});
+    }
+    else
+    {
+        // The init value is manually applied through the write functor in exclusive-scan-by-segment and we always pass
+        // __no_init_value to the transform scan call. This is because init handling must occur on a per-segment basis
+        // and functions differently than the typical scan init which is only applied once in a single location.
+        _WrappedInitType __wrapped_init{{0, __init.__value}};
```

Comment:
Perhaps this makes it more complicated, but do we actually need the wrapped init type to be a `_PackedFlagValueType`?
It seems like in practice in the `WriteOp`, we only ever use the `_ValueType` element, so can we just provide that?
Of course, we can pass `oneapi::dpl::unseq_backend::__no_init_value<_PackedFlagValueType>{}` for the init of the transform scan operation, but it seems like for the segmented part we could have an init type of  `_ValueType` (without the packed tuple and even without the init value wrapper.

I think we could could remove the `conditional_t` part above as well in favor of types / values tailored to the constexpr branches.

Reply:
This will change `_WriteOp`, but actually since inclusive scan never has or uses an init anyway, you can just use `__write_scan_by_seg<__is_inclusive, _ValueType, _BinaryOperator>` for both inclusive and exclusive (and I believe it should be able to move up to join the rest of the using definitions).

Reply:
Yes, this works and makes things simpler. We can just pass `__init` directly to `_WriteOp` in both cases allowing us to remove the constexpr conditional between inclusive and exclusive scan-by-segment.